ClientCore/ProfanityFilter.cs

```diff
@@ -1,90 +1,97 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Text.RegularExpressions;
+using System.Linq;
+using System.Text;
+using System.Diagnostics;
 
 namespace ClientCore
 {
     public class ProfanityFilter
     {
-        public IList<string> CensoredWords { get; private set; }
+        private const char CENSOR_CHAR = '*';
+        private readonly Regex _combinedRegex;
 
         /// <summary>
         /// Creates a new profanity filter with a default set of censored words.
         /// </summary>
-        public ProfanityFilter()
-        {
-            CensoredWords = new List<string>()
+        /// <param name="extraWords">Any extra words to be considered profane.</param>
+        /// 
+        public ProfanityFilter(IEnumerable<string> extraWords = null)
+        {        
+            var defaultWords = new[]
             {
-                "cunt*",
-                "*nigg*",
-                "paki*",
-                "shit",
-                "fuck*",
-                "admin*",
-                "allahu*",
-                "akbar",
-                "twat",
-                "cock",
-                "pussy",
-                "hitler*",
-                "anal"
+                "fuck", "shit", "cunt", "nigger", "nigga", "niggr", "cock",
+                "hitler", "pussy", "akbar", "allahu", "paki", "twat"
             };
-        }
+            var allWords = defaultWords.Concat(extraWords ?? Enumerable.Empty<string>());
 
-        public ProfanityFilter(IEnumerable<string> censoredWords)
-        {
-            if (censoredWords == null)
-                throw new ArgumentNullException("censoredWords");
-            CensoredWords = new List<string>(censoredWords);
+            //for each bad word
+            //      for each letter in the bad word,
+            //          match l33tspeak variations (ex: sh!t)
+            //          match repeating letters    (ex: sshhhhi!iiit)
+            //          ignore hyphens/underscores/whitespace after the letter (ex: s_h_!_t)
+            // join into one big, ugly pattern
+
+            var patterns = allWords
+                .Select(word => string.Join(
+                    @"[-_\s]*",  //ignore hyphens/underscores/whitespace after the letter (ex: s_h !_t)
+                    word.Select(c => 
+                    {
+                        switch (char.ToLower(c))
+                        {
+                            case 'a': return "[aA4@]+"; //in between [] are leetspeak variations.
+                            case 'b': return "[bB8]+";  //and a + matches repeating letters
+                            case 'e': return "[eE3]+";
+                            case 'g': return "[gG69]+";
+                            case 'i': return "[iI1!]+";
+                            case 'o': return "[oO0]+";
+                            case 's': return "[sS5$]+";
+                            case 't': return "[tT7]+";
+                            case 'z': return "[zZ2]+";
```

Comment:
*A person with a Russian keyboard walks into a bar and writes "пigger"*

In other words, I think this isn't a robust method and only works for English language. What I would do is to have lookalike characters placed in a set, for example. You'd have a list of sets and just query each one of them, and build a regex out of this. Maybe it's also a good idea to add localization support here too somehow, because otherwise it would be an uphill battle not worth taking.